main.go

```diff
@@ -182,50 +182,45 @@ func main() {
 					continue
 				}
 
-				taintsToRemove := []int{}
+				// only proceed if all the tolerated daemonset pods on the node are ready
+				allPodsReady := false
 				for _, ipod := range pods {
 					pod, ok := ipod.(*apiv1.Pod)
 					if !ok {
 						continue
 					}
-					if !apiv1pod.IsPodReady(pod) {
-						continue
-					}
 					controller := metav1.GetControllerOfNoCopy(pod)
 					if controller == nil || controller.Kind != "DaemonSet" {
 						continue
 					}
-
-					// TODO: build map of taints with original index to avoid inner loop
-					value := pod.ObjectMeta.Namespace + "." + controller.Name
-					for i, taint := range node.Spec.Taints {
-						if taint.Key != TaintNodeDaemonSetNotReady || taint.Value != value {
-							continue
+					toleratedPod := false
+					for _, toleration := range pod.Spec.Tolerations {
+						if toleration.Key == TaintNodeDaemonSetNotReady {
+							toleratedPod = true
+							break
 						}
-
-						taintsToRemove = append(taintsToRemove, i)
-						// continue to remove any additional taints with the same key
 					}
+					if !toleratedPod {
+						continue
+					}
+					if !apiv1pod.IsPodReady(pod) {
+						allPodsReady = false
+						break
+					}
+					allPodsReady = true
```

Comment:
Is the idea here that we need at least one ready pod here to remove the taint? Might be more intentional to track the count of ready, tolerated, daemonset pods. The boolean logic here works but is a little funky.

Reply:
Curious if we should wait some interval on brand new nodes before checking. Since we're not explicitly waiting on a set of pods now, there could be a race where a daemonset pod has not yet been scheduled but we already see 1 other ready pod.